DR-1895 - Generic retry firestore transaction

snf2ye opened the pull request:
- By using the new runTransactionWithRetry method, we're giving a more substantial wait & retry setup the transactions in FireStoreDirectoryDao.java.  [This is the change that I think is fixing the larger bulk file ingests and snapshot create]
- Converts the runTransactionWithRetry to accept/return generic types so that we can use it for more transactions
- Removes "AbortedException" from retry list for single transactions. For the following exception, I don't think we want to immediately retry the transaction - it would be better to do a full step retry in this case. 
`AbortedException: io.grpc.StatusRuntimeException: ABORTED: Aborted due to cross-transaction contention. This occurs when multiple transactions attempt to access the same data, requiring Firestore to abort at least one in order to enforce serializability."`
- Add "FIRESTORE_RETRIES" config value so that I could set the retry count lower during tests. This helped fix one of the connected tests.

Testing:
Successfully ran the CreateSnapshot testrunner test that ingests 1000 files and creates a snapshot.

snf2ye committed:
wip

snf2ye committed:
revert testing setup

snf2ye committed:
first pass

snf2ye committed:
wip

snf2ye committed:
convert to generic

snf2ye committed:
bump back retry val

snf2ye committed:
fix bugs

snf2ye committed:
handle diff exceptions, add back per test

tlangs commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -192,43 +192,24 @@ public void deleteDirectoryEntriesFromCollection(Firestore firestore, String col
+                    if (docSnap == null) {
+                        return null;
+                    }
+                    return docSnap.toObject(FireStoreDirectoryEntry.class);
```
This can be condensed to 
```
return Optional.ofNullable(docSnap).map(d -> docSnap.toObject(FireStoreDirectoryEntry.class))
                        .orElse(null);
```

tlangs commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -238,13 +219,22 @@ public FireStoreDirectoryEntry retrieveByPath(
+                    if (docSnap == null) {
+                        return null;
+                    }
+                    return docSnap.toObject(FireStoreDirectoryEntry.class);
```
Same here:
```
return Optional.ofNullable(docSnap).map(d -> docSnap.toObject(FireStoreDirectoryEntry.class))
                        .orElse(null);
```

tlangs commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java:
```diff
@@ -250,52 +255,55 @@ String computeCrc32c(String input) {
             throwable instanceof StatusRuntimeException) {
 
+            return true;
+        } else if (isBatch && throwable instanceof AbortedException) {
```
This clause can be added to the above `if` statement since they both result in returning `true`

tlangs commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java:
```diff
@@ -250,52 +255,55 @@ String computeCrc32c(String input) {
-    public void runTransactionWithRetry(Firestore firestore,
+    public <T> T runTransactionWithRetry(Firestore firestore,
                                         FirestoreFunction firestoreFunction,
+                                        Class<T> transactionType,
```
It looks like `transactionType` isn't used in this method.

pshapiro4broad committed:
- make the runTransactionWithRetry method parameterized on the transaction return type (#970)

- convert functional subclasses to lambda expressions
- use the firestore transaction function type instead of defining our own

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -104,8 +102,7 @@ public void createDirectoryEntry(
         fireStoreUtils.runTransactionWithRetry(firestore,
-            (xn) -> {
-
+            (Transaction.Function<Void>) xn -> {
```
I was looking at this more and it turns out you can omit the `(Transaction.Function<Void>)` cast, if you don't mind the inferred return type being `Object` instead of `Void`. Java infers the type from the type of the `return` expression.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -192,43 +189,20 @@ public void deleteDirectoryEntriesFromCollection(Firestore firestore, String col
-        }
-        throw new FileSystemExecutionException(transactionOp + " failed - no more retries");
-    }
+    private static final int SLEEP_BASE_MILLISECONDS = 1000;
```
It's a minor thing but I would use a duration type that lets me express the value in fewer digits. In this case it seems like seconds would be a more natural duration type to use.
```suggestion
    private static final int SLEEP_BASE_SECONDS = 1;
```

snf2ye replied:
I think we've discussed this before! https://github.com/DataBiosphere/jade-data-repo/pull/853/files#r584991592 
I'll go ahead and switch to seconds. Hopefully that will prevent another bug like you caught in one of your other comments.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -238,13 +212,18 @@ public FireStoreDirectoryEntry retrieveByPath(
+                if (docSnap == null) {
+                    return null;
+                }
+                return docSnap.toObject(FireStoreDirectoryEntry.class);
```
Does it make sense to have this code inside the retry block? That is, is there a case where it can fail and need to be retried? If not, I think it would read better to compute `DocumentSnapshot` inside the retry, and have code after the retry that converts it to a `FireStoreDirectoryEntry`

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -627,16 +606,18 @@ private DocumentSnapshot lookupByPathNoXn(
                 lastException = fireStoreUtils.handleExecutionException(ex, "lookupByPathNoXn");
             }
-            TimeUnit.SECONDS.sleep(LOOKUP_WAIT_SECONDS);
+            final long retryWait = SLEEP_BASE_MILLISECONDS * Double.valueOf(Math.pow(2.5, retryNum)).longValue();
```
Does this produce a different result than casting to `long`? Personally I find the cast easier to read.
```suggestion
            final long retryWait = (long) (SLEEP_BASE_MILLISECONDS * Math.pow(2.5, retryNum));
```

snf2ye replied:
With the following results, I think either would work just fine!
I ran this test:
  
    private static final int SLEEP_BASE_MILLISECONDS = 1000;

    @Test
    public void testRetryTime(){
        for(int retryNum = 0; retryNum < 10; retryNum++) {
            final long retryWait = SLEEP_BASE_MILLISECONDS * Double.valueOf(Math.pow(2.5, retryNum)).longValue();
            final long retryWait2 = (long) (SLEEP_BASE_MILLISECONDS * Math.pow(2.5, retryNum));
            logger.info("{} vs. {}", retryWait, retryWait2);
        }
    }
`

![image](https://user-images.githubusercontent.com/13254229/122982470-e95a4980-d368-11eb-9ac4-6f9cf940a9d5.png)

snf2ye replied:
When I convert to seconds, they produce the same result. So, I'll go with the one you suggested!

pshapiro4broad replied:
The results make sense, the original code is truncating `double` to `long` before the multiplication, and the new code truncates afterwards. You could keep the original behavior if you want by writing `SLEEP_BASE_MILLISECONDS * (long) Math.pow(2.5, retryNum)`

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -627,16 +606,18 @@ private DocumentSnapshot lookupByPathNoXn(
             }
-            TimeUnit.SECONDS.sleep(LOOKUP_WAIT_SECONDS);
+            final long retryWait = SLEEP_BASE_MILLISECONDS * Double.valueOf(Math.pow(2.5, retryNum)).longValue();
+            TimeUnit.SECONDS.sleep(retryWait);
```
Is this a bug? Your sleep base duration is milliseconds but this is using seconds.
```suggestion
            TimeUnit.MILLISECONDS.sleep(retryWait);
```

snf2ye replied:
Yes, this is a bug! Thank you for catching it!

snf2ye committed:
code review changes

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -627,16 +606,18 @@ private DocumentSnapshot lookupByPathNoXn(
+            TimeUnit.SECONDS.sleep(retryWait);
         }
+        logger.error("Ran out of retries - lookupPathNoXn");
         throw lastException;
```
Note that this will NPE if someone sets `FireStoreutils.firestoreRetries` to zero. You may want to add a check in the `FireStoreUtils` constructor and force it to `1` if it's less than `1`.

snf2ye replied:
Thanks - I'm wondering where the best place is to put this check. Would GoogleResourceConfiguration.java be better than in FirestoreUtils? Might as well get it correctly from the source?

pshapiro4broad replied:
It looks like there's not much logic in our configuration classes. I can see the check going into either place. I think I like `FirestoreUtils` better because that will put the check closer to the reason for the check.

Looking at the code in `FirestoreUtils`, it looks like it's written so that at least one attempt will occur even if `firestoreRetries` is `0`. So maybe this code should be changed to work that way too, so no minimum value is needed.

snf2ye replied:
Ok, I've changed the FirestoreDirectoryDao so that it always runs at least once, even if the max is zero.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java:
```diff
@@ -66,21 +70,21 @@ RuntimeException handleExecutionException(Throwable ex, String op) {
         if (throwable instanceof FirestoreException) {
             FirestoreException fex = (FirestoreException) throwable;
-            String msg = "FirestoreUtils.handleExecutionException - Retrying firestore exception: " + fex;
+            String msg = "[" + op + "] FirestoreUtils.handleExecutionException - Retrying firestore exception: " + fex;
```
You could use `{}` here too instead of string concatenation. I don't have a preference one way or the other but it's odd to see it done inconsistently in a single method.

snf2ye replied:
Can you use that when defining a regular string? I thought we could only do that for logger messages.

pshapiro4broad replied:
Oh, I see, the message is used for both the logger and the exception. No, in that case I don't think there's a better way.

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreUtils.java:
```diff
@@ -250,52 +255,45 @@ String computeCrc32c(String input) {
             } catch (Exception ex) {
                 final long retryWait = SLEEP_BASE_MILLISECONDS * Double.valueOf(Math.pow(2.5, retry)).longValue();
-                if (FireStoreUtils.shouldRetry(ex) && retry < firestoreRetries) {
+                if (FireStoreUtils.shouldRetry(ex, false) && retry < firestoreRetries) {
```
A minor thing, I would do the `<` operation first since `&&` short circuits and I'd bet `<` is faster than `FireStoreUtils.shouldRetry()`. 
```suggestion
                if (retry < firestoreRetries && FireStoreUtils.shouldRetry(ex, false)) {
```

pshapiro4broad reviewed (changes requested):
A few minor things and one possible bug

snf2ye committed:
pass of code review changes

snf2ye committed:
move parse out of firestore retry

snf2ye committed:
add check for config

snf2ye committed:
code review change

snf2ye committed:
remove another void

pshapiro4broad commented on src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDirectoryDao.java:
```diff
@@ -192,44 +190,18 @@ public void deleteDirectoryEntriesFromCollection(Firestore firestore, String col
+            xn -> lookupByFileId(firestore, collectionId, fileId, xn),
             "retrieveById", " file id: " + fileId);
+
+        return Optional.ofNullable(docSnap).map(d -> docSnap.toObject(FireStoreDirectoryEntry.class))
+            .orElse(null);
```
OK, I was going to say we shouldn't be doing this, but then I realized that there's a better solution. Instead of making an `Optional` here, `FirestoreUtils. runTransactionWithRetry()` could return `Optional<T>` instead of `<T>`.

And really, _this_ method should return `Optional<FireStoreDirectoryEntry>`. But that's probably something for another PR.

pshapiro4broad replied:
"shouldn't be doing this"  == making an `Optional` out of an object just to test `!= null`

snf2ye replied:
Sounds like a great idea! My first pass at implementing this didn't go so well, so I'd like to put this in a follow on PR where I can test this change out more.

pshapiro4broad reviewed (approved):
This looks good. It would be nice to refactor the retry code more but that can wait for a future PR. I think making the return type `Optional` would be a nice improvement, but the code is fine without it too.

snf2ye committed:
code review changes

snf2ye committed:
correct comments

snf2ye committed:
revert changes to lookupByPathNoXn, fix one test

snf2ye committed:
fix connected tests

snf2ye committed:
Revert "fix connected tests"

This reverts commit 8f3bb397d4a603c9f3287910d9a74a1eafd8d383.

snf2ye committed:
still throw on AbortedException

snf2ye committed:
cleanup

snf2ye committed:
fix tests

snf2ye committed:
remove aborted exception again

snf2ye committed:
code cleanup
